cmd/ocp-build-data-enforcer/main.go

```diff
@@ -0,0 +1,304 @@
+	flag.StringVar(&o.majorMinor.minor, "minor", "6", "The minor version to target")
+	flag.Parse()
+	return o
+}
+func main() {
+	opts := gatherOptions()
+	opts.majorMinor.major = "4"
+
+	configsUnverified, err := gatherAllOCPImageConfigs(opts.ocpBuildDataRepoDir, opts.majorMinor)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to gather all ocp image configs")
+	}
+
+	streamMap, err := readStreamMap(opts.ocpBuildDataRepoDir, opts.majorMinor)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to read streamMap")
+	}
+
+	groupYAML, err := readGroupYAML(opts.ocpBuildDataRepoDir, opts.majorMinor)
+	if err != nil {
+		logrus.WithError(err).Fatal("Faild to read groupYAML")
+	}
+
+	var errs []error
+	var configs []ocpImageConfig
+	for _, cfg := range configsUnverified {
+		if err := cfg.validate(); err != nil {
+			errs = append(errs, fmt.Errorf("error validating %s: %w", cfg.SourceFileName, err))
+			continue
+		}
+		if err := dereferenceConfig(&cfg, streamMap, groupYAML); err != nil {
+			errs = append(errs, fmt.Errorf("failed dereferencing config for %s: %w", cfg.SourceFileName, err))
+			continue
+		}
+		configs = append(configs, cfg)
+	}
+
+	errGroup := &errgroup.Group{}
+	for idx := range configs {
+		idx := idx
+		errGroup.Go(func() error {
+			processDockerfile(configs[idx])
+			return nil
+		})
+	}
+	if err := errGroup.Wait(); err != nil {
+		logrus.WithError(err).Fatal("Processing failed")
+	}
+
+	if err := utilerrors.NewAggregate(errs); err != nil {
+		for _, err := range err.Errors() {
+			logrus.WithError(err).Error("Encountered error")
+		}
+		logrus.Fatal("Encountered errors")
+	}
+	logrus.Infof("Processed %d configs", len(configs))
+}
+
+// TODO (alvaroaleman): This is incomplete
+func dereferenceConfig(config *ocpImageConfig, streamMap streamMap, groupYAML groupYAML) error {
+	if replacement, hasReplacement := streamMap[config.From.Stream]; hasReplacement {
+		config.From.Stream = replacement.UpstreamImage
+	}
+	for blder := range config.From.Builder {
+		if replacement, hasReplacement := streamMap[config.From.Builder[blder].Stream]; hasReplacement {
+			config.From.Builder[blder].Stream = replacement.UpstreamImage
+		}
+	}
+	if config.Content.Source.Alias != "" {
+		if _, hasReplacement := groupYAML.Sources[config.Content.Source.Alias]; !hasReplacement {
+			return fmt.Errorf("groups.yaml has no replacement for alias %s", config.Content.Source.Alias)
+		}
+		config.Content.Source.Git = &ocpImageConfigSourceGit{}
+		*config.Content.Source.Git = groupYAML.Sources[config.Content.Source.Alias]
+	}
+
+	return nil
+}
+
+func processDockerfile(config ocpImageConfig) {
+	orgRepo := config.orgRepo()
+	log := logrus.WithField("file", config.SourceFileName).WithField("org/repo", orgRepo)
+	split := strings.Split(orgRepo, "/")
+	if n := len(split); n != 2 {
+		log.Errorf("splitting orgRepo didn't yield 2 but %d results", n)
+		return
+	}
+	if split[0] == "openshift-priv" {
```

Comment:
So it's not that we need to handle updating Dockerfiles in private forks for public repos, but for e.g. `openshift/origin` I think there is not and will not be an entry in `ocp-build-data`. In that case we need to find the actual upstream we need to update by dereferencing (with the publicize config?). We should check that we understand this workflow with Justin.

Reply:
They have a mapping for that which I will be using in the other PR: https://github.com/openshift/ocp-build-data/blob/e8c46fac4007e65ed13338ec0c29633119a5d417/group.yml#L74